README.md
# WordPress Review Me Library
This library provides developers to ask end user to review their product (theme / plugin)

[![Latest Stable Version](https://poser.pugx.org/varunsridharan/wp-review-me/version)](https://packagist.org/packages/varunsridharan/wp-review-me)
[![Total Downloads](https://poser.pugx.org/varunsridharan/wp-review-me/downloads)](https://packagist.org/packages/varunsridharan/wp-review-me)
[![Latest Unstable Version](https://poser.pugx.org/varunsridharan/wp-review-me/v/unstable)](//packagist.org/packages/varunsridharan/wp-review-me)
[![License](https://poser.pugx.org/varunsridharan/wp-review-me/license)](https://packagist.org/packages/varunsridharan/wp-review-me)
[![composer.lock available](https://poser.pugx.org/varunsridharan/wp-review-me/composerlock)](https://packagist.org/packages/varunsridharan/wp-review-me)


## Installation
The preferred way to install this extension is through [Composer](http://getcomposer.org/download/).

To install **WP_Review_Me library**, simply:

    $ composer require Varunsridharan/WP_Review_Me

The previous command will only install the necessary files, if you prefer to **download the entire source code** you can use:

    $ composer require Varunsridharan/WP_Review_Me --prefer-source

You can also **clone the complete repository** with Git:

    $ git clone https://github.com/varunsridharan/wp-review-me.git

Or **install it manually**:

[Download WP_Review_Me.php](https://raw.githubusercontent.com/varunsridharan/wp-review-me/master/class-wp-review-me.php):

    $ wget https://raw.githubusercontent.com/varunsridharan/wp-review-me/master/class-wp-review-me.php



## Options
| Option | Notes |
| ------ | ----- |
| `days_after` | Value should be in numbers |
| `slug` | Plugin Slug (Required) |
| `site` | Supported Sites : `wordpress` / `themeforest` / `codecanyon` |
| `item_id` | Item ID required if site is set-to `themeforest` / `codecanyon` |
| `type` | Set `plugin` or `theme` if site is set-to **wordpress** |
| `rating` | Set Int Val if site is set-to **wordpress** |
| `message` | Custom Message To Show In Admin |
| `link_label` | Review Button Label |
| `review_link` | You can override auto link generation by providing a link here |
| `notice_callback` | If you are using any admin notice handler / custom Library then you can provide a function to call when notice should be shown |



## Usage 

#### Plugin / Theme From Wordpress.org
```php
$review_instance = new Varunsridharan\WordPress\Review_Me(array(
    'days_after' => 2,
    'slug' => 'your-plugin',
    'type' => 'plugin', # Use theme if you are using it in a theme
    'site' => 'wordpress',
    'rating' => 3 # review link will have a default count of rating 3 which sets rating to 3 in review form
));
```

#### Plugin From Codecanyon.net
```php
$review_instance = new Varunsridharan\WordPress\Review_Me(array(
    'days_after' => 2,
    'slug' => 'your-plugin',
    'site' => 'codecanyon',
    'item_id' => 'xxxxxxxx'
));
```

#### Theme From themeforest.net
```php
$review_instance = new Varunsridharan\WordPress\Review_Me(array(
    'days_after' => 2,
    'slug' => 'your-plugin',
    'site' => 'themeforest',
    'item_id' => 'xxxxxxxx'
));
```


#### Custom Notice Callback
```php
/**
 * $review_me is a instance of VS_WP_Review_Me Class
 */
function your_plugin_review_notice_add($review_me){
    // use your custom notice handle to add / display the notice
}

$review_instance = new Varunsridharan\WordPress\Review_Me(array(
    'days_after' => 2,
    'slug' => 'your-plugin',
    'type' => 'plugin', # Use theme if you are using it in a theme
    'site' => 'wordpress',
    'notice_callback' => 'your_plugin_review_notice_add'
));
```

## Sample Output
[![View 1](https://raw.githubusercontent.com/varunsridharan/wp-review-me/master/screenshot-1.jpg)](https://raw.githubusercontent.com/varunsridharan/wp-review-me/master/screenshot-1.jpg)

[![View 2](https://raw.githubusercontent.com/varunsridharan/wp-review-me/master/screenshot-2.jpg)](https://raw.githubusercontent.com/varunsridharan/wp-review-me/master/screenshot-2.jpg)


---

<!-- START common-footer.mustache  -->
## 📝 Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

[Checkout CHANGELOG.md](https://github.com/varunsridharan/wp-review-me/blob/main/CHANGELOG.md)


## 🤝 Contributing
If you would like to help, please take a look at the list of [issues](https://github.com/varunsridharan/wp-review-me/issues/).


## 📜  License & Conduct
- [**GNU General Public License v3.0**](https://github.com/varunsridharan/wp-review-me/blob/main/LICENSE) © [Varun Sridharan](website)
- [Code of Conduct](https://github.com/varunsridharan/.github/blob/main/CODE_OF_CONDUCT.md)


## 📣 Feedback
- ⭐ This repository if this project helped you! :wink:
- Create An [🔧 Issue](https://github.com/varunsridharan/wp-review-me/issues/) if you need help / found a bug


## 💰 Sponsor
[I][twitter] fell in love with open-source in 2013 and there has been no looking back since! You can read more about me [here][website].
If you, or your company, use any of my projects or like what I’m doing, kindly consider backing me. I'm in this for the long run.

- ☕ How about we get to know each other over coffee? Buy me a cup for just [**$9.99**][buymeacoffee]
- ☕️☕️ How about buying me just 2 cups of coffee each month? You can do that for as little as [**$9.99**][buymeacoffee]
- 🔰         We love bettering open-source projects. Support 1-hour of open-source maintenance for [**$24.99 one-time?**][paypal]
- 🚀         Love open-source tools? Me too! How about supporting one hour of open-source development for just [**$49.99 one-time ?**][paypal]

<!-- Personl Links -->
[paypal]: https://sva.onl/paypal
[buymeacoffee]: https://sva.onl/buymeacoffee
[twitter]: https://sva.onl/twitter/
[website]: https://sva.onl/website/


## Connect & Say 👋
- **Follow** me on [👨‍💻 Github][github] and stay updated on free and open-source software
- **Follow** me on [🐦 Twitter][twitter] to get updates on my latest open source projects
- **Message** me on [📠 Telegram][telegram]
- **Follow** my pet on [Instagram][sofythelabrador] for some _dog-tastic_ updates!

<!-- Personl Links -->
[sofythelabrador]: https://www.instagram.com/sofythelabrador/
[github]: https://sva.onl/github/
[twitter]: https://sva.onl/twitter/
[telegram]: https://sva.onl/telegram/


---

<p align="center">
<i>Built With ♥ By <a href="https://sva.onl/twitter"  target="_blank" rel="noopener noreferrer">Varun Sridharan</a> <a href="https://en.wikipedia.org/wiki/India">
   <img src="https://cdn.svarun.dev/flag-india.jpg" width="20px"/></a> </i> <br/><br/>
   <img src="https://cdn.svarun.dev/codeispoetry.png"/>
</p>

---


<!-- END common-footer.mustache  -->


CHANGELOG.md
# 📝  Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## 1.6 - 11/04/2021
### Changed
* Codeclean up

### Removed
* `defined( 'ABSPATH' ) || exit;` To avoid script crashing while running PHPCS

## 1.5 - 07/05/2019
### Added
* `class_exists` check in php

## 1.4 - 06/05/2019
### Fixed
* Issue with [wordpress.org](https://wordpress.org) link generation

## 1.3 - 14/01/2019
### Fixed
* Issue With Autoloader

## 1.2.1 - 20/12/2018
### Changed
* `composer.json` With PSR4 Standards Autoloader

## 1.2 - 20/12/2018
### Changed
* `composer.json` With PSR4 Standards Autoloader

## 1.1 - 20/12/2018
### Added
* `composer.json` Auto load Support

## 1.0 - 17/12/2018
### First Release


<!--
## Unreleased

## 1.0 - 01/02/2020
### Added

### Changed

### Deprecated

### Removed

### Fixed

### Security

-->


class-wp-review-me.php
<?php
/**
 * WordPress Review Me Library
 * This library provides developers to ask end user to review their product (theme / plugin)
 *
 * @author    Varun Sridharan <varunsridharan23@gmail.com>
 * @copyright 2018 Varun Sridharan
 * @license   GPLV3 Or Greater
 */

namespace Varunsridharan\WordPress;

if ( ! defined( 'WPINC' ) ) {
	die;
}

/**
 * Class Review_Me
 *
 * @package Varunsridharan\WordPress
 * @author Varun Sridharan <varunsridharan23@gmail.com>
 * @since 1.0
 */
class Review_Me {
	/**
	 * Library Version
	 *
	 * @var string
	 */
	public $version = '1.0';

	/**
	 * linkid
	 *
	 * @var
	 */
	public $linkid;

	/**
	 * key
	 *
	 * @var string
	 */
	protected $key;

	/**
	 * link_id
	 *
	 * @var null|string
	 */
	protected $link_id = null;

	/**
	 * op
	 *
	 * @var array
	 */
	protected $op = array();

	/**
	 * Review_Me constructor.
	 *
	 * @param array $options
	 */
	public function __construct( $options = array() ) {
		if ( \is_admin() || defined( 'DOING_AJAX' ) ) {
			$this->op      = \wp_parse_args( $options, $this->get_defaults() );
			$this->key     = 'vswprm_' . substr( md5( $this->op['slug'] ), 0, 20 );
			$this->link_id = 'vswprm-review-link-' . $this->key;

			if ( \get_option( $this->key . 'isdone', false ) === false ) {
				$this->maybe_prompt();
			}

			\add_action( 'wp_ajax_vswprm_clicked_review', array( $this, 'dismiss_notice' ) );
		}
	}

	/**
	 * returns defaults args
	 *
	 * @return array
	 */
	public function get_defaults() {
		return array(
			'days_after'      => 10,
			'notice_callback' => false,
			'type'            => '',
			'slug'            => '',
			'rating'          => 5,
			'message'         => \sprintf( \esc_html__( 'Hey! It&#039;s been a little while that you&#039;ve been using this product. You might not realize it, but user reviews are such a great help to us. We would be so grateful if you could take a minute to leave a review on WordPress.org. Many thanks in advance :)', 'wp-review-me' ) ),
			'link_label'      => \esc_html__( 'Click here to leave your review', 'wp-review-me' ),
			'review_link'     => false,
			'item_id'         => false,
			'site'            => 'wordpress',
		);
	}

	/**
	 * Checks And Shows Review Message
	 */
	protected function maybe_prompt() {
		if ( ! $this->is_time() ) {
			return;
		}

		\add_action( 'admin_footer', array( $this, 'script' ) );
		if ( false !== $this->op['notice_callback'] ) {
			call_user_func_array( $this->op['notice_callback'], array( &$this ) );
		} else {
			\add_action( 'admin_notices', array( $this, 'add_notice' ) );
		}
	}

	/**
	 * Check if it is time to ask for a review
	 *
	 * @since 1.0
	 * @return bool
	 */
	protected function is_time() {
		$installed = \get_option( $this->key, false );

		if ( false === $installed || 0 === $installed ) {
			$this->setup_date();
			$installed = time();
		}

		if ( $installed + ( $this->op['days_after'] * 86400 ) > time() ) {
			return false;
		}
		return true;
	}

	/**
	 * Save the current date as the installation date
	 *
	 * @since 1.0
	 * @return void
	 */
	protected function setup_date() {
		\update_option( $this->key, time() );
	}

	/**
	 * Display Admin Notice
	 */
	public function add_notice() {
		echo '<div class="updated success is-dismissible"><p>' . $this->get_message() . '</p></div>';
	}

	/**
	 * Get the review prompt message
	 *
	 * @since 1.0
	 * @return string
	 */
	public function get_message() {
		$message = $this->op['message'];
		$link    = $this->get_review_link_tag();
		$message = $message . ' ' . $link;
		return \wp_kses_post( $message );
	}

	/**
	 * Get the complete link tag
	 *
	 * @since 1.0
	 * @return string
	 */
	protected function get_review_link_tag() {
		$link  = $this->get_review_link();
		$label = $this->op['link_label'];
		return "<a href='$link' target='_blank' id='$this->link_id'>$label</a>";
	}

	/**
	 * Get the review link
	 *
	 * @since 1.0
	 * @return string
	 */
	protected function get_review_link() {
		if ( false === $this->op['review_link'] ) {
			$link = '';
			if ( 'wordpress' === $this->op['site'] ) {
				$link = 'https://wordpress.org/support/';

				switch ( $this->op['type'] ) {
					case 'theme':
						$link .= 'theme/';
						break;
					case 'plugin':
						$link .= 'plugin/';
						break;
				}

				$link .= $this->op['type'] . '/reviews';

				$link = \add_query_arg( 'rate', $this->op['rating'], $link ) . \esc_url( $link . '#new-post' );
			} elseif ( 'codecanyon' === $this->op['site'] ) {
				$link = 'https://codecanyon.net/item/x/reviews/' . $this->op['item_id'] . '#rating-' . $this->op['item_id'];
			} elseif ( 'themeforest' === $this->op['site'] ) {
				$link = 'https://themeforest.net/item/x/reviews/' . $this->op['item_id'] . '#rating-' . $this->op['item_id'];
			}

			return $link;
		}

		return $this->op['review_link'];
	}

	/**
	 * Echo the JS script in the admin footer
	 *
	 * @since 1.0
	 * @return void
	 */
	public function script() { ?>

		<script>
			jQuery( document ).ready( function( $ ) {
				$( '#<?php echo $this->link_id; ?>' ).on( 'click', wrmDismiss );

				function wrmDismiss() {
					var data = {
						action: 'vswprm_clicked_review',
						id: '<?php echo $this->link_id; ?>'
					};
					jQuery.ajax( {
						type: 'POST',
						url: ajaxurl,
						data: data,
					} );
				}
			} );
		</script>

	<?php }

	/**
	 * Dismiss the notice when the review link is clicked
	 *
	 * @since 1.0
	 * @return void
	 */
	public function dismiss_notice() {
		if ( empty( $_POST ) ) {
			die();
		}
		if ( ! isset( $_POST['id'] ) ) {
			die();
		}
		$id = \sanitize_text_field( $_POST['id'] );
		if ( $id !== $this->link_id ) {
			die();
		}

		\update_option( $this->key . 'isdone', 'yes' );
		die();
	}
}


src/Review_Me.php
<?php
/**
 * WordPress Review Me Library
 * This library provides developers to ask end user to review their product (theme / plugin)
 *
 * @author    Varun Sridharan <varunsridharan23@gmail.com>
 * @copyright 2018 Varun Sridharan
 * @license   GPLV3 Or Greater
 */

namespace Varunsridharan\WordPress;


if ( ! class_exists( '\Varunsridharan\WordPress\Review_Me' ) ) {
	/**
	 * Class Review_Me
	 *
	 * @package Varunsridharan\WordPress
	 * @author Varun Sridharan <varunsridharan23@gmail.com>
	 * @since 1.0
	 */
	class Review_Me {
		/**
		 * Library Version
		 *
		 * @var string
		 */
		public $version = '1.4';

		/**
		 * linkid
		 *
		 * @var
		 */
		public $linkid;

		/**
		 * key
		 *
		 * @var string
		 */
		protected $key;

		/**
		 * link_id
		 *
		 * @var null|string
		 */
		protected $link_id = null;

		/**
		 * op
		 *
		 * @var array
		 */
		protected $op = array();

		/**
		 * Review_Me constructor.
		 *
		 * @param array $options
		 */
		public function __construct( $options = array() ) {
			if ( \is_admin() || defined( 'DOING_AJAX' ) ) {
				$this->op      = \wp_parse_args( $options, $this->get_defaults() );
				$this->key     = 'vswprm_' . substr( md5( $this->op['slug'] ), 0, 20 );
				$this->link_id = 'vswprm-review-link-' . $this->key;

				if ( \get_option( $this->key . 'isdone', false ) === false ) {
					$this->maybe_prompt();
				}

				\add_action( 'wp_ajax_vswprm_clicked_review', array( $this, 'dismiss_notice' ) );
			}
		}

		/**
		 * returns defaults args
		 *
		 * @return array
		 */
		public function get_defaults() {
			return array(
				'days_after'      => 10,
				'notice_callback' => false,
				'type'            => '',
				'slug'            => '',
				'rating'          => 5,
				'message'         => \sprintf( \esc_html__( 'Hey! It&#039;s been a little while that you&#039;ve been using this product. You might not realize it, but user reviews are such a great help to us. We would be so grateful if you could take a minute to leave a review on WordPress.org. Many thanks in advance :)', 'wp-review-me' ) ),
				'link_label'      => \esc_html__( 'Click here to leave your review', 'wp-review-me' ),
				'review_link'     => false,
				'item_id'         => false,
				'site'            => 'wordpress',
			);
		}

		/**
		 * Checks And Shows Review Message
		 */
		protected function maybe_prompt() {
			if ( ! $this->is_time() ) {
				return;
			}

			\add_action( 'admin_footer', array( $this, 'script' ) );
			if ( false !== $this->op['notice_callback'] ) {
				call_user_func_array( $this->op['notice_callback'], array( &$this ) );
			} else {
				\add_action( 'admin_notices', array( $this, 'add_notice' ) );
			}
		}

		/**
		 * Check if it is time to ask for a review
		 *
		 * @return bool
		 * @since 1.0
		 */
		protected function is_time() {
			$installed = \get_option( $this->key, false );

			if ( false === $installed || 0 === $installed ) {
				$this->setup_date();
				$installed = time();
			}

			if ( $installed + ( $this->op['days_after'] * 86400 ) > time() ) {
				return false;
			}
			return true;
		}

		/**
		 * Save the current date as the installation date
		 *
		 * @return void
		 * @since 1.0
		 */
		protected function setup_date() {
			\update_option( $this->key, time() );
		}

		/**
		 * Display Admin Notice
		 */
		public function add_notice() {
			echo '<div class="updated success is-dismissible"><p>' . $this->get_message() . '</p></div>';
		}

		/**
		 * Get the review prompt message
		 *
		 * @return string
		 * @since 1.0
		 */
		public function get_message() {
			$message = $this->op['message'];
			$link    = $this->get_review_link_tag();
			$message = $message . ' ' . $link;
			return \wp_kses_post( $message );
		}

		/**
		 * Get the complete link tag
		 *
		 * @return string
		 * @since 1.0
		 */
		protected function get_review_link_tag() {
			$link  = $this->get_review_link();
			$label = $this->op['link_label'];
			return "<a href='$link' target='_blank' id='$this->link_id'>$label</a>";
		}

		/**
		 * Get the review link
		 *
		 * @return string
		 * @since 1.0
		 */
		protected function get_review_link() {
			if ( false === $this->op['review_link'] ) {
				$link = '';
				if ( 'wordpress' === $this->op['site'] ) {
					$link = 'https://wordpress.org/support/';

					switch ( $this->op['type'] ) {
						case 'theme':
							$link .= 'theme/';
							break;
						case 'plugin':
							$link .= 'plugin/';
							break;
					}

					$link .= $this->op['slug'] . '/reviews';

					$link = \add_query_arg( 'rate', $this->op['rating'], $link ) . '#new-post';
				} elseif ( 'codecanyon' === $this->op['site'] ) {
					$link = 'https://codecanyon.net/item/x/reviews/' . $this->op['item_id'] . '#rating-' . $this->op['item_id'];
				} elseif ( 'themeforest' === $this->op['site'] ) {
					$link = 'https://themeforest.net/item/x/reviews/' . $this->op['item_id'] . '#rating-' . $this->op['item_id'];
				}

				return $link;
			}

			return $this->op['review_link'];
		}

		/**
		 * Echo the JS script in the admin footer
		 *
		 * @return void
		 * @since 1.0
		 */
		public function script() { ?>

			<script>
				jQuery( document ).ready( function( $ ) {
					$( '#<?php echo $this->link_id; ?>' ).on( 'click', wrmDismiss );

					function wrmDismiss() {
						var data = {
							action: 'vswprm_clicked_review',
							id: '<?php echo $this->link_id; ?>'
						};
						jQuery.ajax( {
							type: 'POST',
							url: ajaxurl,
							data: data,
						} );
					}
				} );
			</script>

		<?php }

		/**
		 * Dismiss the notice when the review link is clicked
		 *
		 * @return void
		 * @since 1.0
		 */
		public function dismiss_notice() {
			if ( empty( $_POST ) ) {
				die();
			}
			if ( ! isset( $_POST['id'] ) ) {
				die();
			}
			$id = \sanitize_text_field( $_POST['id'] );
			if ( $id !== $this->link_id ) {
				die();
			}

			\update_option( $this->key . 'isdone', 'yes' );
			die();
		}
	}
}